### ISSUE TITLE
After computer wakes up websocket fails to recover

### ISSUE THREAD

**Issue by lc-ikrumov** [Atmosphere/atmosphere#2211]
If a client computer that has an open websocket connection sleeps for a few minutes and then wakes up, the websocket connection is not re-established and fails with many errors "Failed to execute 'send'" (see attached log). Is this something intended? How can I make sure the websocket remains alive?

(Sorry if I should be creating a 'atmosphere-javascript' issue, I am not sure if the problem is in the server-side or not)

Client console log:
[console.txt](https://github.com/Atmosphere/atmosphere/files/459133/console.txt)

The onError callback is called, which I believe is called only on a non-recoverable error. That is why when I receive such an error I downgrade the connection to my own polling mechanism (unsubscribe from Atmosphere and rely on a legacy mechanism as a safety measure). But a sleeping computer is supposed to be a recoverable state, the websocket should be reconnected in some way, and this downgrade should not happen.

I have seen the same behaviour when the client temporarily disconnects from the network. 

Here's what I'm reproducing it with:

**Client:**
Atmosphere Client v2.3.2
Client OS: OS X El Capitan 10.11.5
Client Browser: Chrome 52.0.2743.116 (64-bit)
Transports: websocket (main), long-polling (fallback)

**Client Config:**
contentType: 'application/json',
transport: 'websocket',
trackMessageLength: true,
enableXDR: true,
rewriteURL: true,
shared: true,
timeout: 0,
connectTimeout: 30000,
reconnectInterval: 5000,
fallbackTransport: 'long-polling',
reconnectOnServerError: false

**Server:**
Atmosphere Server - Runtime Native v2.4.5, Compat Tomcat v2.0.1, JBossWeb v2.0.1
Web server: Tomcat 7.0.70 with APR connector (latest APR and native libs)
Server OS distro: RedHat 6.3

**Server Config:**
org.atmosphere.cpr.Broadcaster.scanClassPath=false
org.atmosphere.cpr.broadcaster.shareableThreadPool=true
org.atmosphere.cpr.broadcasterLifeCyclePolicy=EMPTY
org.atmosphere.useWebSocketAndServlet3=true
org.atmosphere.websocket.messageContentType=application/json
org.atmosphere.websocket.writeTimeout=30000
org.atmosphere.useNative=false
org.atmosphere.interceptor.HeartbeatInterceptor.heartbeatFrequencyInSeconds=60
org.atmosphere.interceptor.HeartbeatInterceptor.clientHeartbeatFrequencyInSeconds=10
org.atmosphere.cpr.broadcaster.maxProcessingThreads=25
org.atmosphere.cpr.broadcaster.maxAsyncWriteThreads=25
org.atmosphere.pool.BoundedApachePoolableProvider.size=200
org.atmosphere.cpr.broadcasterFactory=org.atmosphere.pool.PoolableBroadcasterFactory
org.atmosphere.pool.poolableProvider=org.atmosphere.pool.BoundedApachePoolableProvider
org.atmosphere.cpr.recoverFromDestroyedBroadcaster=true
org.atmosphere.websocket.bannedVersion=-1

**Interceptors:**
CorsInterceptor, AtmosphereResourceLifecycleInterceptor, TrackMessageSizeInterceptor, 
HeartbeatInterceptor, SuspendTrackerInterceptor, AtmosphereResourceStateRecovery

**Filters:**
WebSocketHandshakeFilter

**Comment by lc-ikrumov** [comment]
I found out what the issue is:
When the computer goes to sleep it behaves the same as if it is disconnected from the internet. Atmosphere client detects that it can no longer communicate to the server and triggers the onError() callback. In my callback I (mistakenly) issued an unsubscribe() request on the connection (not knowing that Atmosphere already closed the websocket before triggering onError()), which requires the server to be reachable - since it is not, an error is thrown again and onError() gets triggered, leading to an infinite loop of onError() calls.

This eventually crashes the browser tab but if the internet connectivity is re-established fast enough the browser does not crash, the infinite onError() loop stops eventually and my code downgrades to legacy polling mechanism (which happens in onError()).

So, I fixed it by removing the line that unsubscribes (closes the socket) in onError. The socket now recovers after internet connectivity loss.
